# Coding Exercises (Part 1)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas Basics (DataFrame Basics I)

### Exercise 1: Importing and Inspecting your own Dataset

Before we can work with the Pandas Library, we need to __import Pandas__ (by convention as __pd__). <br>
1. __Import__ Pandas. __Fill in the gaps__!

In [1]:
import pandas as pd

Now we want to import the __Cars Dataset__ that is stored in the CSV-File __cars.csv__. <br>
2. __Import__ the Cars Dataset and __store__ the Pandas DataFrame in the variable __cars__! __Fill in the gaps__!

In [2]:
cars = pd.read_csv("cars.csv")

#### Some more details on the Columns:

- __mpg__: Measures __fuel efficiency__ in terms of miles per gallon. __Higher mpg__ means __higher efficiency__
- __cylinders__: Number of cylinders in the car
- __displacement__: Technical feature (engine)
- __horsepower__: Unit of measurement for engine power 
- __weight__: weight in lbs
- __acceleration__: time to accelerate from 0 to 60 mph (sec.)
- __model_year__: self explanatory
- __origin__: origin of the manufacturer/car
- __name__: model name

3. __Inspect__ the __first 10 Rows__ of the DataFrame cars! __Fill in the gaps__! What is the __name of the very first car__ in the first row (index label/pos = 0)?

In [3]:
cars.head(10)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   
5  15.0          8         429.0       198.0    4341          10.0   
6  14.0          8         454.0       220.0    4354           9.0   
7  14.0          8         440.0       215.0    4312           8.5   
8  14.0          8         455.0       225.0    4425          10.0   
9  15.0          8         390.0       190.0    3850           8.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino  
5          70    usa           ford galaxie 500  
6          70    usa           chevrolet impala  
7          70    usa          plymouth fury iii  
8          70    usa           pontiac catalina  
9          70    usa         amc ambassador dpl

It´s a ... Chevrolet Chevelle Malibu!

4. Also inspect the __last 5 Rows__! __How many cars/rows__ do we have in the Dataset (mind the range index starting from 0!)?

In [4]:
cars.tail()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin             name  
393          82     usa  ford mustang gl  
394          82  europe        vw pickup  
395          82     usa    dodge rampage  
396          82     usa      ford ranger  
397          82     usa   chevrolet s-10

We have ... 398 Cars!

5. Get some __meta information__ on our DataFrame! In which __Column__ do we have __missing/NaN__ Values?

In [5]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
mpg             398 non-null float64
cylinders       398 non-null int64
displacement    398 non-null float64
horsepower      392 non-null float64
weight          398 non-null int64
acceleration    398 non-null float64
model_year      398 non-null int64
origin          398 non-null object
name            398 non-null object
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


We have 6 missing/NaN Values in the column ... horsepower!

6. Let´s get some __summary statistics__ on our DataFrame! What is the __maximum__ value in the column __horsepower__?

In [6]:
cars.describe()

mpg   cylinders  displacement  horsepower       weight  \
count  398.000000  398.000000    398.000000  392.000000   398.000000   
mean    23.514573    5.454774    193.425879  104.469388  2970.424623   
std      7.815984    1.701004    104.269838   38.491160   846.841774   
min      9.000000    3.000000     68.000000   46.000000  1613.000000   
25%     17.500000    4.000000    104.250000   75.000000  2223.750000   
50%     23.000000    4.000000    148.500000   93.500000  2803.500000   
75%     29.000000    8.000000    262.000000  126.000000  3608.000000   
max     46.600000    8.000000    455.000000  230.000000  5140.000000   

       acceleration  model_year  
count    398.000000  398.000000  
mean      15.568090   76.010050  
std        2.757689    3.697627  
min        8.000000   70.000000  
25%       13.825000   73.000000  
50%       15.500000   76.000000  
75%       17.175000   79.000000  
max       24.800000   82.000000

The maximum value in the column horsepower is ... 230!

-------------------------------

### Exercise 2: Indexing and Slicing

7. __Import Pandas__ (as pd), <br>
__Import__ the Cars Dataset (__cars.csv__) and save the DataFrame in the variable __cars__ and<br>
Inspect the __first 5 Rows__!

In [7]:
import pandas as pd

In [8]:
cars = pd.read_csv("cars.csv")

In [9]:
cars.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

8. Select the column __"name"__ with the __attribute__ notation! What is the __name__ of the car with __index label 8__?

In [10]:
cars.name

0              chevrolet chevelle malibu
1                      buick skylark 320
2                     plymouth satellite
3                          amc rebel sst
4                            ford torino
5                       ford galaxie 500
6                       chevrolet impala
7                      plymouth fury iii
8                       pontiac catalina
9                     amc ambassador dpl
10                   dodge challenger se
11                    plymouth 'cuda 340
12                 chevrolet monte carlo
13               buick estate wagon (sw)
14                 toyota corona mark ii
15                       plymouth duster
16                            amc hornet
17                         ford maverick
18                          datsun pl510
19                  vw 1131 deluxe sedan
20                           peugeot 504
21                           audi 100 ls
22                              saab 99e
23                              bmw 2002
24              

The name is ... pontiac catalina!

9. Apply also the __second alternative__ to select the column __"name"__!

In [11]:
cars["name"]

0              chevrolet chevelle malibu
1                      buick skylark 320
2                     plymouth satellite
3                          amc rebel sst
4                            ford torino
5                       ford galaxie 500
6                       chevrolet impala
7                      plymouth fury iii
8                       pontiac catalina
9                     amc ambassador dpl
10                   dodge challenger se
11                    plymouth 'cuda 340
12                 chevrolet monte carlo
13               buick estate wagon (sw)
14                 toyota corona mark ii
15                       plymouth duster
16                            amc hornet
17                         ford maverick
18                          datsun pl510
19                  vw 1131 deluxe sedan
20                           peugeot 504
21                           audi 100 ls
22                              saab 99e
23                              bmw 2002
24              

10. __Select__ the car/row at __index position 100__ with the __iloc__ operator (all columns)! How many __cylinders__ does the car have?

In [12]:
cars.iloc[100]

mpg                        18
cylinders                   6
displacement              250
horsepower                 88
weight                   3021
acceleration             16.5
model_year                 73
origin                    usa
name            ford maverick
Name: 100, dtype: object

The car has ... 6 cylinders!

11. Select the car/row at __index position 200__ with the __iloc__ operator (and only the column __"name"__)! <br>
What is the __name__ of the car?

In [13]:
cars.iloc[200, 8]

'ford granada ghia'

It´s a ... ford granada ghia!

12. Select the __last 10 rows__ and the columns __mpg, horsepower, origin__ and __name__ with the __iloc__ operator! __Fill in the Gaps__! <br>
What is the __mpg__ of the __vw pickup__?

In [14]:
cars.iloc[-10:, [0, 3, 7, 8]]

mpg  horsepower  origin                        name
388  26.0        92.0     usa  chrysler lebaron medallion
389  22.0       112.0     usa              ford granada l
390  32.0        96.0   japan            toyota celica gt
391  36.0        84.0     usa           dodge charger 2.2
392  27.0        90.0     usa            chevrolet camaro
393  27.0        86.0     usa             ford mustang gl
394  44.0        52.0  europe                   vw pickup
395  32.0        84.0     usa               dodge rampage
396  28.0        79.0     usa                 ford ranger
397  31.0        82.0     usa              chevrolet s-10

The mpg is ... 44.0!

13. __Import__ the __Cars Dataset__ one more time (save in the variable __cars__) and set the column __"name"__ as the __index__! <br>
__Fill in the Gaps__!

In [15]:
cars = pd.read_csv("cars.csv", index_col = "name")

In [16]:
#run the cell and inspect!
cars.head()

mpg  cylinders  displacement  horsepower  weight  \
name                                                                           
chevrolet chevelle malibu  18.0          8         307.0       130.0    3504   
buick skylark 320          15.0          8         350.0       165.0    3693   
plymouth satellite         18.0          8         318.0       150.0    3436   
amc rebel sst              16.0          8         304.0       150.0    3433   
ford torino                17.0          8         302.0       140.0    3449   

                           acceleration  model_year origin  
name                                                        
chevrolet chevelle malibu          12.0          70    usa  
buick skylark 320                  11.5          70    usa  
plymouth satellite                 11.0          70    usa  
amc rebel sst                      12.0          70    usa  
ford torino                        10.5          70    usa

In [17]:
#run the cell and inspect!
cars.info()

<class 'pandas.core.frame.DataFrame'>
Index: 398 entries, chevrolet chevelle malibu to chevrolet s-10
Data columns (total 8 columns):
mpg             398 non-null float64
cylinders       398 non-null int64
displacement    398 non-null float64
horsepower      392 non-null float64
weight          398 non-null int64
acceleration    398 non-null float64
model_year      398 non-null int64
origin          398 non-null object
dtypes: float64(4), int64(3), object(1)
memory usage: 28.0+ KB


14. __Inspect__ the new Index of the DataFrame by calling the __index__ attribute!

In [18]:
cars.index

Index(['chevrolet chevelle malibu', 'buick skylark 320', 'plymouth satellite',
       'amc rebel sst', 'ford torino', 'ford galaxie 500', 'chevrolet impala',
       'plymouth fury iii', 'pontiac catalina', 'amc ambassador dpl',
       ...
       'chrysler lebaron medallion', 'ford granada l', 'toyota celica gt',
       'dodge charger 2.2', 'chevrolet camaro', 'ford mustang gl', 'vw pickup',
       'dodge rampage', 'ford ranger', 'chevrolet s-10'],
      dtype='object', name='name', length=398)

15. Select the __row__ with the index label __"ford ranger"__ (use the __loc__ operator)! What is the __weight__?

In [19]:
cars.loc["ford ranger"]

mpg               28
cylinders          4
displacement     120
horsepower        79
weight          2625
acceleration    18.6
model_year        82
origin           usa
Name: ford ranger, dtype: object

The weight is ... 2625!

16. Select the __row__ with the index label __"ford torino"__ and only the columns __"model_year"__ and __"origin"__! <br>
What´s the __model year__?

In [20]:
cars.loc["ford torino", ["model_year", "origin"]]

model_year     70
origin        usa
Name: ford torino, dtype: object

The model year is ... 70!

17. Select the columns __"mpg"__ and __"weight"__ (all rows) with the __loc__ operator!

In [21]:
cars.loc[:, ["mpg", "weight"]]

mpg  weight
name                                           
chevrolet chevelle malibu          18.0    3504
buick skylark 320                  15.0    3693
plymouth satellite                 18.0    3436
amc rebel sst                      16.0    3433
ford torino                        17.0    3449
ford galaxie 500                   15.0    4341
chevrolet impala                   14.0    4354
plymouth fury iii                  14.0    4312
pontiac catalina                   14.0    4425
amc ambassador dpl                 15.0    3850
dodge challenger se                15.0    3563
plymouth 'cuda 340                 14.0    3609
chevrolet monte carlo              15.0    3761
buick estate wagon (sw)            14.0    3086
toyota corona mark ii              24.0    2372
plymouth duster                    22.0    2833
amc hornet                         18.0    2774
ford maverick                      21.0    2587
datsun pl510                       27.0    2130
vw 1131 deluxe sedan               26.0    1835
peugeot 504                        25.0    2672
audi 100 ls                        24.0    2430
saab 99e                           25.0    2375
bmw 2002                           26.0    2234
amc gremlin                        21.0    2648
ford f250                          10.0    4615
chevrolet c20                      10.0    4376
dodge d200                         11.0    4382
ih 1200d                            9.0    4732
datsun pl510                       27.0    2130
...                                 ...     ...
chevrolet cavalier wagon           27.0    2640
chevrolet cavalier 2-door          34.0    2395
pontiac j2000 se hatchback         31.0    2575
dodge aries se                     29.0    2525
pontiac phoenix                    27.0    2735
ford fairmont futura               24.0    2865
amc concord dl                     23.0    3035
vw rabbit l                        36.0    1980
mazda glc custom l                 37.0    2025
mazda glc custom                   31.0    1970
plymouth horizon miser             38.0    2125
mercury lynx l                     36.0    2125
nissan stanza xe                   36.0    2160
honda accord                       36.0    2205
toyota corolla                     34.0    2245
honda civic                        38.0    1965
honda civic (auto)                 32.0    1965
datsun 310 gx                      38.0    1995
buick century limited              25.0    2945
oldsmobile cutlass ciera (diesel)  38.0    3015
chrysler lebaron medallion         26.0    2585
ford granada l                     22.0    2835
toyota celica gt                   32.0    2665
dodge charger 2.2                  36.0    2370
chevrolet camaro                   27.0    2950
ford mustang gl                    27.0    2790
vw pickup                          44.0    2130
dodge rampage                      32.0    2295
ford ranger                        28.0    2625
chevrolet s-10                     31.0    2720

[398 rows x 2 columns]

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Hints (Spoiler!)

1. pandas as pd

2. Pass the filename in quotation marks to pd.read_csv()

3. head() method with n = 10

4. tail() method

5. info() method

6. describe() method

7. pandas as pd, pd.read_csv("filename"), head()

8. cars.---

9. cars["---"]

10. cars.iloc[---]

11. cars.iloc[row index position, column index position]; "name" column is at index position ... 8

12. The last 10 rows = from index position -10 till the end (:); Count the column index positions of the Columns


13. parameter index_col

14. cars.---

15. cars.loc["row label"]

16. cars.loc[---, [---, ---]]

17.  cars.loc[ : , [---, ---]]